app/code/community/Apruve/ApruvePayment/Model/PaymentMethod.php

```diff
@@ -24,161 +24,207 @@
+			}
+			if (!$this->canUseForCountry($billingCountry)) {
+				Mage::throwException($this->_getHelper()->__('Selected payment type is not allowed for billing country.'));
+			}
+			return $this;
+		} elseif (!$this->getInfoInstance()->getAdditionalInformation('aprt')) {
+			Mage::throwException('Something is going wrong, try again to post order with apruve.');
+		}
+	}
+
+	/**
+	 * Get token and create transaction
+	 * @param Varien_Object $payment
+	 * @param float $amount
+	 * @return Mage_Payment_Model_Abstract|void
+	 * @throws Mage_Core_Exception
+	 */
+	public function authorize(Varien_Object $payment, $amount)
+	{
+		parent::authorize($payment, $amount);
+		Mage::helper('apruvepayment')->logException('Authorize...');
+
+		$additionalInformation = $payment->getAdditionalInformation();
+		$token = $additionalInformation['aprt'];
+
+		/** @var Mage_Sales_Model_Order $order */
+		$order = $payment->getOrder();
+
+		/** @var Apruve_ApruvePayment_Model_Api_Rest_Order $orderApi */
+		$orderApi = Mage::getModel('apruvepayment/api_rest_order');
+
+
+		$updateResult = $orderApi->updateOrder($token, $order);
+		if (!$updateResult || !$updateResult['success']) {
+			Mage::throwException('Couldn\'t update order in Apruve.');
+		}
+
+		$payment->setTransactionId($updateResult['response']['id']);
+		$payment->setIsTransactionClosed(0);
+
+		return $this;
+	}
+
+	/**
+	 * Captures a payment
+	 *
+	 * @param   Varien_Object $payment
+	 * @return  bool
+	 * @throws  Mage_Core_Exception
+	 */
+	public function capture(Varien_Object $payment, $amount)
+	{
+		parent::capture($payment, $amount);
+		Mage::helper('apruvepayment')->logException('Capture...');
+
+		if ($amount <= 0) {
+			Mage::throwException(Mage::helper('paygate')->__('Invalid amount for capture.'));
+		}
+
+		$payment->setSkipTransactionCreation(true);
+		return $this;
+	}
+
+
+	/**
+	 * Create a refund
+	 *
+	 * @param   Varien_Object $payment
+	 * @return  bool
+	 * @throws  Mage_Core_Exception
+	 */
+	public function refund(Varien_Object $payment, $amount){
+		Mage::helper('apruvepayment')->logException('Refund...');
+
+		// Get Magento Order
+		$order = $payment->getOrder();
+		$apruveEntity = Mage::getModel('apruvepayment/entity')->loadByOrderId($order->getIncrementId(), 'magento_id');
+		$apruveOrderId = $apruveEntity->getApruveId();
+
+		// Get API objects
+		$orderApi = Mage::getModel('apruvepayment/api_rest_order');
+		$invoiceApi = Mage::getModel('apruvepayment/api_rest_invoice');
+
+		// User Order to Get Apruve invoices
+		$invoices = $orderApi->getInvoices($apruveOrderId);
+
+		$validInvoiceIds = [];
+		$totalAmountCents = 0;
+
+		foreach ($invoices as $invoice){
+			$validInvoiceIds[] = $invoice['id'];
+			$totalAmountCents += $invoice['amount_cents'];
+		}
+
+		if ($totalAmountCents >= ($amount * 100)  && !empty($validInvoiceIds)){
+			$result = $invoiceApi->refundInvoice($validInvoiceIds[0], $amount);
```

Comment:
I found a scenario when this will not work. Imagine there are two invoices on the order, both paid by the buyer. Then I create a credit memo for the first invoice in Magento. Then I attempt to create a credit memo for the second. Based on this logic it will attempt to create a second credit memo for the first invoice which will not work because of the validations in the API. 

We somehow need to determine exactly which invoice the credit memo is being applied to.